Softeq.XToolkit.PushNotifications.Droid/Services/DroidPushNotificationsConsumer.cs

```diff
@@ -0,0 +1,182 @@
+        /// <param name="notificationsSettings">Provides settings for notification construction.</param>
+        /// <param name="pushNotificationsParser">Parses remote messages into common format.</param>
+        /// <param name="pushTokenStorageService">Stores push notification token and it's status.</param>
+        /// <param name="pushNotificationsHandler">Handles received push notifications events.</param>
+        /// <param name="remotePushNotificationsService">Propagates push notification token to remote server.</param>
+        /// <param name="showForegroundNotificationsInSystemOptions">Defines how push notification should be presented.</param>
+        /// <param name="logManager">Provides logging.</param>
+        public DroidPushNotificationsConsumer(
+            INotificationsSettingsProvider notificationsSettings,
+            IPushNotificationsParser pushNotificationsParser,
+            IPushTokenStorageService pushTokenStorageService,
+            IPushNotificationsHandler pushNotificationsHandler,
+            IRemotePushNotificationsService remotePushNotificationsService,
+            ForegroundNotificationOptions showForegroundNotificationsInSystemOptions,
+            ILogManager logManager)
+        {
+            _pushTokenStorageService = pushTokenStorageService;
+            _pushNotificationsHandler = pushNotificationsHandler;
+            _remotePushNotificationsService = remotePushNotificationsService;
+            _showForegroundNotificationsInSystemOptions = showForegroundNotificationsInSystemOptions;
+            _pushNotificationsParser = pushNotificationsParser;
+            _logger = logManager.GetLogger<DroidPushNotificationsConsumer>();
+
+            NotificationsHelper.Init(notificationsSettings);
+
+            _lifecycleObserver = new AppLifecycleObserver();
+            ProcessLifecycleOwner.Get().Lifecycle.AddObserver(_lifecycleObserver);
+        }
+
+        /// <inheritdoc />
+        public bool TryHandleNotification(RemoteMessage message)
+        {
+            if (!_pushNotificationsParser.TryParse(message, out var parsedNotification))
+            {
+                return false;
+            }
+
+            OnMessageReceivedInternal(parsedNotification, _lifecycleObserver.IsForegrounded);
+
+            if (!parsedNotification.IsSilent && (_showForegroundNotificationsInSystemOptions.ShouldShow() || !_lifecycleObserver.IsForegrounded))
+            {
+                NotificationsHelper.CreateNotification(Application.Context, parsedNotification, message.Data);
+            }
+
+            return true;
+        }
+
+        /// <inheritdoc />
+        public bool TryHandlePushNotificationIntent(Intent intent)
+        {
+            if (!_pushNotificationsParser.TryParse(intent.Extras, out var parsedNotification))
+            {
+                return false;
+            }
+
+            _pushNotificationsHandler.HandlePushNotificationTapped(parsedNotification);
+
+            return true;
```

Comment:
It returns `false` if notification cannot be parsed.

Reply:
Sorry, I mean `HandlePushNotificationTapped` method result

Reply:
Are you suggesting that `HandlePushNotificationTapped` should return boolean value, indicating if push-notification tap has been handled or not? I'm not sure it would be a good idea, in current implementation at least: if notification was parsed, but was not handled - should anyone else try to handle it?

The general idea, in my mind, is that push notification always originates from a specific provider, and should be handled by corresponding consumer ONLY. That's why we need some kind of definitive indicator of wether push notification belongs to a consumer or not - I decided to use notification parsing as this definitive indicator (specifically, format of the notification's payload). It means that if notification is parsed (has expected format), then it does belong to this consumer, and this consumer should handle this notification and not pass it to anyone else, even if parsed notification cannot be handled for some reason (for instance, because of some internal error).